src/main/java/io/r2dbc/pool/ConnectionPool.java

```diff
@@ -163,27 +164,50 @@ public void close() {
         dispose();
     }
 
+    /**
+     * Dispose this {@link ConnectionPool} in non-blocking flow.
+     * <p>
+     * When multiple errors occurred during dispose flow, they are added as
+     * suppressed errors onto the first error.
+     *
+     * @return a Mono triggering the shutdown of the pool once subscribed.
+     */
+    public Mono<Void> disposeLater() {
+        List<Throwable> errors = new ArrayList<>();
+        return Flux.fromIterable(this.destroyHandlers)
+            .flatMap(Mono::fromRunnable)
+            .concatWith(this.connectionPool.disposeLater())
+            .onErrorContinue((throwable, o) -> {
+                errors.add(throwable);
+            })
+            .then(Mono.defer(() -> {
+                if (errors.isEmpty()) {
+                    return Mono.empty();
+                }
+
+                Throwable rootError = errors.get(0);
+                if (errors.size() == 1) {
+                    return Mono.error(rootError);
+                }
+
+                errors.subList(1, errors.size()).forEach(rootError::addSuppressed);
+
+                return Mono.error(rootError);
+            }));
+    }
+
     @Override
     public void dispose() {
 
-        RuntimeException exception = null;
-
-        for (Runnable destroyHandler : this.destroyHandlers) {
-            try {
-                destroyHandler.run();
-            } catch (RuntimeException e) {
-                if (exception == null) {
-                    exception = e;
-                } else {
-                    exception.addSuppressed(e);
-                }
-            }
-        }
+        AtomicReference<Throwable> errorHolder = new AtomicReference<>();
 
-        this.connectionPool.dispose();
+        disposeLater()
```

Comment:
Let's stick with Pool's disposal strategy to make this method non-blocking and non-error aware.

Otherwise, we would need to call `.block()` here as `.subscribe()` is non-blocking and this method could terminate before the actual disposal is finished.

Reply:
So, you mean keeping the existing implementation for `dispose()` method which performs `destroyHandlers`,  calls `this.connectionPool.dispose()`, and adds suppressed exceptions; so that this `dispose()` method is the imperative way of disposing this connection pool??

Reply:
I would keep the code changes as proposed by this PR and add here just `disposeLater().subscribe()`.